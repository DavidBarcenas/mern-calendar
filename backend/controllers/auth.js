const { response } = require('express')
const bcrypt = require('bcryptjs')
const User = require('../models/User')
const { generateJWT } = require('../utils/jwt')

const createUser = async (req, res = response) => {
  const {email, pwd} = req.body
  try {
    let user = await User.findOne({ email })

    if(user) {
      return res.status(400).json({
        ok: false,
        msg: 'El correo ya está en uso'
      })
    }
    user = new User(req.body);

    const salt = bcrypt.genSaltSync()
    user.pwd = bcrypt.hashSync(pwd, salt)

    await user.save()
    const token = await generateJWT(user.id, user.name)

    res.status(201).json({
      ok: true,
      uid: user.id,
      name: user.name,
      token
    })
  } catch (error) {
    res.status(500).json({
      ok: false,
      msg: 'Contacte a su administrador'
    })
  }
}

const loginUser = async (req, res = response) => {
  const { email, pwd} = req.body

  try {
    const user = await User.findOne({ email })

    if(!user) {
      return res.status(400).json({
        ok: false,
        msg: 'El usuario no existe'
      })
    }

    const validPwd = bcrypt.compareSync(pwd, user.pwd)

    if(!validPwd) {
      return res.status(400).json({
        ok: false,
        msg: 'El correo o la contraseña es incorrecto'
      })
    }

    // JWT
    const token = await generateJWT(user.id, user.name)

    res.status(200).json({
      ok: true,
      uid: user.id,
      name: user.name,
      token
    })
    
  } catch (error) {
    res.status(500).json({
      ok: false,
      msg: 'Contacte a su administrador'
    })
  }
}

const reValidateJWT = async (req, res = response) => {
  const token = await generateJWT(req.uid, req.name)

  res.json({
    ok: true,
    token
  })
}

module.exports = {
  createUser,
  loginUser,
  reValidateJWT
}