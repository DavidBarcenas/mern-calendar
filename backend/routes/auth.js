/**
 * User Routes / Auth
 * host + /api/auth
 */

const express = require('express')
const { check } = require('express-validator')
const { createUser, loginUser, reValidateJWT } = require('../controllers/auth')
const { validateFields } = require('../middlewares/validate-fields')

const router = express.Router()

router.post('/', [
  check('email', 'Email is required and must be valid').isEmail(),
  check('pwd', 'Password must have a minimum of 6 characters').isLength({min: 6}),
  validateFields
], loginUser)

router.post(
  '/new', [
    check('name', 'Name is required').not().isEmpty(),
    check('email', 'Email is required and must be valid').isEmail(),
    check('pwd', 'Password must have a minimum of 6 characters').isLength({min: 6}),
    validateFields
  ], createUser)
  
router.get(
  '/renew', reValidateJWT)

module.exports = router