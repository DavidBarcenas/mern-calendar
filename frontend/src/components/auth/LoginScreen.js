import React, { useState } from 'react'
import TextField from '@material-ui/core/TextField';
import { Button, useMediaQuery } from '@material-ui/core';

export const LoginScreen = () => {

  const [loginAnimation, setLoginAnimation] = useState({
    signIn: false,
    animate: false
  })

  const mq = useMediaQuery('(min-width: 768px)');

  const handleAnimation = (isLogin) => {
    setLoginAnimation({ 
      ...loginAnimation, 
      signIn: isLogin,
    })
    setTimeout(() => {
      setLoginAnimation({ 
          signIn: isLogin,
          animate: isLogin,
        })
    }, 200);
  }
  
  const { signIn, animate } = loginAnimation

  return (
    <div className='login__wrapper'>
      <div className="login__box" style={{transform: signIn && mq ? 'translateX(100%)' : 'translateX(0)'}}>
        {
          !animate ? 
            <div className="login__signin">
              <h1>CalendarApp</h1>
              <h2 className="login__title">Iniciar Sesión</h2>
              <form noValidate autoComplete="off">
                  <TextField label="Correo electrónico" variant="outlined" type="email" />
                  <TextField label="Contraseña" type='password' variant="outlined" />
                  <p className="forgout">¿Olvidaste tu contraseña?</p>
                  <Button variant="outlined" color="primary"> INGRESAR </Button>
              </form>
            </div> : 
            <div className="login__signin"> 
              <h2 className="login__title">Registrarse</h2>
              <form noValidate autoComplete="off">
                <TextField label="Nombre" type="text" variant="outlined" />
                <TextField label="Correo electrònico" type="email" variant="outlined" />
                <TextField label="Contraseña" type="password" variant="outlined" />
                <TextField label="Repetir contraseña" type="password" variant="outlined" />
                <Button variant="outlined" color="primary"> Crear cuenta </Button>
              </form>
            </div>
        }
      </div>

      <div className='login__overlay' style={{transform: signIn && mq ? 'translateX(-100%)' : 'translateX(0)'}}>
        {
          !animate ? 
          <div>
              <h2 className="login__title">¡Hola Amigo!</h2>
              <p>Ingresa tus datos personales y comienza tu viaje con nosotros.</p>
              <Button variant="outlined" color="primary" onClick={() => handleAnimation(true)}> REGISTRARSE </Button>
            </div>
          : 
          <div>
              <h2 className="login__title">¡Bienvenido de nuevo!</h2>
              <p>Para mantenerte conectado con nosotros, ingresa con tus datos de acceso.</p>
              <Button variant="outlined" color="primary" onClick={() => handleAnimation(false)}> INICIAR SESIÓN </Button>
            </div>
        }
        </div>
    </div>
  )
}
