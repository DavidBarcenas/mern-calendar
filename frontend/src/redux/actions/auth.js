import { tokenlessFetch } from "../../utils/fetch"
import { types } from "../types/types"

export const startLogin = (email, pwd) => {
  return async (dispatch) => {
    const resp = await tokenlessFetch('auth', {email, pwd}, 'POST')
    const body = await resp.json()

    if(body.ok) {
      localStorage.setItem('token', body.token)
      localStorage.setItem('token-init', new Date().getTime())

      dispatch(login({
        uid: body.uid,
        name: body.name
      }))
    }
  }
}

const login = (user) => ({
  type: types.authLogin,
  payload: user
})