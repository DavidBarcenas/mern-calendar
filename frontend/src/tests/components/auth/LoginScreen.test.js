import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import React from 'react'
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk';
import { LoginScreen } from '../../../components/auth/LoginScreen';

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {
}

const store = mockStore(initState)
store.dispatch = jest.fn()

const wrapper = mount(
  <Provider store={store}>
    <LoginScreen />
  </Provider>
)

describe('<LoginSreen /> testing', () => { 
  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  // TODO: test Register
})
