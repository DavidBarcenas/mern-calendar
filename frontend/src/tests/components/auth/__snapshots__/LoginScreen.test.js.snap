// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginSreen /> testing should render correctly 1`] = `
<div
  className="login__wrapper"
>
  <div
    className="login__box"
    style={
      Object {
        "transform": "translateX(0)",
      }
    }
  >
    <div
      className="login__signin"
    >
      <h1>
        CalendarApp
      </h1>
      <h2
        className="login__title"
      >
        Iniciar Sesión
      </h2>
      <form
        autoComplete="off"
        noValidate={true}
        onSubmit={[Function]}
      >
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
            data-shrink={true}
          >
            Correo electrónico
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              name="lEmail"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="email"
              value="davee@gmail.com"
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4"
              >
                <span>
                  Correo electrónico
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
            data-shrink={true}
          >
            Contraseña
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              name="lPwd"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="password"
              value="dave123"
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4"
              >
                <span>
                  Contraseña
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <p
          className="forgout"
        >
          ¿Olvidaste tu contraseña?
        </p>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
          disabled={false}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          <span
            className="MuiButton-label"
          >
             INGRESAR 
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </div>
  <div
    className="login__overlay"
    style={
      Object {
        "transform": "translateX(0)",
      }
    }
  >
    <div>
      <h2
        className="login__title"
      >
        ¡Hola Amigo!
      </h2>
      <p>
        Ingresa tus datos personales y comienza tu viaje con nosotros.
      </p>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
           REGISTRARSE 
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
